### PR TITLE
Stabilize debugger_visualizer

### LINKED ISSUE #95939: Tracking Issue for `debugger_visualizer`
This is a tracking issue for [RFC 3191](https://github.com/rust-lang/rfcs/pull/3191).
The feature gate for the issue is `#![feature(debugger_visualizer)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Implement the RFC
- [ ] Adjust documentation (https://github.com/rust-lang/reference/pull/1335)
- [ ] Stabilization PR (https://github.com/rust-lang/rust/pull/108668)

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

No unresolved questions at this time.

### Implementation history

- [x] Create new attribute `#[debugger_viusalizer]` and add support for embedding NatVis visualizers through this attribute: https://github.com/rust-lang/rust/pull/91779

- [x] Adding support for embedding pretty-printers via new attribute: https://github.com/rust-lang/rust/pull/97028

### PR DESCRIPTION
**Pull request by gibbyfree**
This stabilizes the `debugger_visualizer` attribute (#95939).

* Marks the `debugger_visualizer` feature as `accepted`.
* Marks the `debugger_visualizer` attribute as `ungated`.
* Deletes feature gate test, removes feature gate from other tests.

Closes #95939

### CODE REVIEW

**Comment by oli-obk** [comment]
Do we have any documentation for this feature yet?

The tracking issue links to https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs for the next step.

cc @wesleywiser what's the status of this feature? Any unresolved questions still open?

**Comment by gibbyfree** [comment]
> Do we have any documentation for this feature yet?

I created a PR for documentation here: https://github.com/rust-lang/reference/pull/1335 I'll be removing the 'draft' state shortly.

File: tests/debuginfo/auxiliary/dependency-with-embedded-visualizers.rs [hunk 1]
```diff
@@ -2,7 +2,7 @@
 // ignore-lldb
 // no-prefer-dynamic
 
-#![feature(debugger_visualizer)]
+#![cfg_attr(bootstrap, feature(debugger_visualizer))]
```
Comment (dpaoliello) [inline tests/debuginfo/auxiliary/dependency-with-embedded-visualizers.rs:5]: Don't use `bootstrap` for tests: remove the attribute instead

**Comment by nyurik** [comment]
I keep reading the RFC & the docs, and I keep thinking that we can do much better than natvis.

I have written extensive debugger visualizers back in my C# days, where the visualizations had to be done mostly [as code](https://learn.microsoft.com/en-us/visualstudio/debugger/walkthrough-writing-a-visualizer-in-csharp?view=vs-2022). If we adapt natvis, many modern Rust developers will have a major XML cringe every time they would think about it - and that just means far fewer implementations.

I don't mind XML being stored inside PDBs - out of sight, out of mind, but the developers implementing visualizations should not be aware of it / write it / store it in a repo.  Instead, I believe we should provide absolute minimum needed support in the compiler, and let crates define the needed attributes/macros to generate (natvis?) internally.

Or is this the intention all along?

**Comment by gibbyfree** [comment]
@nyurik Do you see [writing GDB pretty printers](https://sourceware.org/gdb/onlinedocs/gdb/Writing-a-Pretty_002dPrinter.html#Writing-a-Pretty_002dPrinter) as a reasonable alternative to Natvis? This might be the preferred means of embedding visualizations for crate owners less familiar with Natvis.

The intention of this feature is to allow crate owners to specify their own visualizations (Natvis if they'd like to support WinDbg/Visual Studio users, Python if they'd like to support GDB users) to be embedded in their crate's debug binary. This allows crate consumers to visualize a crate's custom types without any manual configuration on their part. This achieves "out of sight, out of mind" from the end-user's perspective.

As far as supporting visualization development via macros or some other form of source embedding, I think this falls under "V2" of debugger_visualizer. This extension would probably require a new set of design discussions. (See https://github.com/rust-lang/rfcs/pull/3191#issuecomment-974442018 for a bit more info.) By stabilizing the current state of debugger_visualizer, the feature can begin adding value to the community while possible extensions are specced out. We have already contributed Natvis definitions to some popular crates (https://github.com/servo/rust-smallvec/pull/286 and https://github.com/Lokathor/tinyvec/pull/167, for example).

**Comment by wesleywiser** [comment]
Hi @gibbyfree, would you mind rebasing this PR and removing the merge commits? We prefer not to have merge commits in the PR history as it makes the final merge history messier. 

Feel free to ping me if `git rebase` gives you trouble!

**Comment by wesleywiser** [comment]
Thanks!

@bors r+ rollup

**Pull request merged**